nomad/csi_endpoint_test.go

```diff
@@ -147,6 +148,140 @@ func TestCSIVolumeEndpoint_Register(t *testing.T) {
 	require.Nil(t, resp2.Volume)
 }
 
+// TestCSIVolumeEndpoint_Claim exercises the VolumeClaim RPC, verifying that claims
+// are honored only if the volume exists, the mode is permitted, and the volume
+// is schedulable according to its count of claims.
+func TestCSIVolumeEndpoint_Claim(t *testing.T) {
+	t.Parallel()
+	srv, shutdown := TestServer(t, func(c *Config) {
+		c.NumSchedulers = 0 // Prevent automatic dequeue
+	})
+	defer shutdown()
+	testutil.WaitForLeader(t, srv.RPC)
+
+	ns := structs.DefaultNamespace
+	state := srv.fsm.State()
+	state.BootstrapACLTokens(1, 0, mock.ACLManagementToken())
+	srv.config.ACLEnabled = true
+	policy := mock.NamespacePolicy(ns, "",
+		[]string{acl.NamespaceCapabilityCSICreateVolume, acl.NamespaceCapabilityCSIAccess})
+	accessToken := mock.CreatePolicyAndToken(t, state, 1001,
+		acl.NamespaceCapabilityCSIAccess, policy)
+	codec := rpcClient(t, srv)
+	id0 := uuid.Generate()
+
+	// Create an initial volume claim request; we expect it to fail
+	// because there's no such volume yet.
+	claimReq := &structs.CSIVolumeClaimRequest{
+		VolumeID:   id0,
+		Allocation: mock.BatchAlloc(),
+		Claim:      structs.CSIVolumeClaimWrite,
+		WriteRequest: structs.WriteRequest{
+			Region:    "global",
+			Namespace: ns,
+			AuthToken: accessToken.SecretID,
+		},
+	}
+	claimResp := &structs.CSIVolumeClaimResponse{}
+	err := msgpackrpc.CallWithCodec(codec, "CSIVolume.Claim", claimReq, claimResp)
+	require.EqualError(t, err, fmt.Sprintf("volume not found: %s", id0),
+		"expected 'volume not found' error because volume hasn't yet been created")
+
+	// Create a client nodes with a plugin
+	node := mock.Node()
+	node.CSINodePlugins = map[string]*structs.CSIInfo{
+		"minnie": {PluginID: "minnie",
+			Healthy:  true,
+			NodeInfo: &structs.CSINodeInfo{},
+		},
+	}
+	plugin := structs.NewCSIPlugin("minnie", 1)
+	plugin.ControllerRequired = false
+	plugin.AddPlugin(node.ID, &structs.CSIInfo{})
+	err = state.UpsertNode(3, node)
+	require.NoError(t, err)
+
+	// Create the volume for the plugin
+	vols := []*structs.CSIVolume{{
+		ID:             id0,
+		Namespace:      "notTheNamespace",
+		PluginID:       "minnie",
```

Comment:
Appreciate the shoutout, the string is a holdover from when these were called `Driver`, thus "minnie" and "adam". Don't want to accidentally create a cargo cult.

Reply:
I was wondering about that! 😆